Allowed to use the same DP many times for sensors, Decode of Base64 encoded payload, Binary payload support, possibility to add new entities for existing device

MaximumSU opened the pull request:
**For sensors:**
* Allowed to use the same DP many times for sensors (this is need if one DP contain more then on values, for example, 3 phase circuit breakers and counters)
* Decode of Base64 encoded payload support
* Binary payload segmentation and decode support
* Round decimals support

**Config flow:**
* Refactoring to allow changes for sensors
* Added possibility to add new entities for existing device without delete them

sibowler committed:
Fix for AttributeError: 'TuyaMessage' object has no attribute 'release' error

rospogrigio committed:
Introduced pytuya with support for 3.4 protocol

rospogrigio committed:
Introduced pytuya with support for 3.4 protocol

rospogrigio committed:
Introduced 3.4 protocol option in config flow

rospogrigio committed:
Introduced 3.4 protocol option in config flow

rospogrigio committed:
Fixed HEARTBEAT command

rospogrigio committed:
Fixed HEARTBEAT command

rospogrigio committed:
Added debugging

rospogrigio committed:
Added debugging

rospogrigio committed:
Fixed requested DPs for type_0d devices

rospogrigio committed:
Fixed requested DPs for type_0d devices

MaximumSU committed:
Allowed to use the same DP many times for sensors, Decode of Base64 encoded payload, Binary payload support, possibility to add new entities for existing device

For sensors:
* Allowed to use the same DP many times for sensors (this is need if one DP contain more then on values, for example, 3 phase circuit breakers and counters)
* Decode of Base64 encoded payload support
* Binary payload segmentation and decode support
* Round decimals support

Config flow:
* Refactoring to allow changes for sensors
* Added possibility to add new entities for existing device without delete them

MaximumSU committed:
Allowed to use the same DP many times for sensors, Decode of Base64 encoded payload, Binary payload support, possibility to add new entities for existing device

For sensors:
* Allowed to use the same DP many times for sensors (this is need if one DP contain more then on values, for example, 3 phase circuit breakers and counters)
* Decode of Base64 encoded payload support
* Binary payload segmentation and decode support
* Round decimals support

Config flow:
* Refactoring to allow changes for sensors
* Added possibility to add new entities for existing device without delete them

rospogrigio committed:
Fixed negotiation and sequence numbers for protocol 3.4

rospogrigio committed:
Fixed negotiation and sequence numbers for protocol 3.4

rospogrigio committed:
Fixed sequence numbering on status update

rospogrigio committed:
Fixed sequence numbering on status update

sibowler committed:
Merge branch 'rospogrigio:master' into fix_semaphore_release_error

MaximumSU committed:
Do not allow to add the sensor with same DP and Bytes Range

It is possible to add many sensors with same DP but not with same DP+Bytes Range combination.

MaximumSU committed:
Do not allow to add the sensor with same DP and Bytes Range

It is possible to add many sensors with same DP but not with same DP+Bytes Range combination.

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Updated README.md and info.md

rospogrigio committed:
Updated README.md and info.md

rospogrigio committed:
Merge branch 'protocol_3.4' of github.com:rospogrigio/localtuya into protocol_3.4

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Fixed tox issues

rospogrigio committed:
Fix for version not being set yet when calling _setup_dispatcher

rospogrigio committed:
Fix for version not being set yet when calling _setup_dispatcher

rospogrigio committed:
Introduced abort in config flow if all entities are deselected when editing a device

rospogrigio committed:
Introduced the possibility to select which devices should have the pytuya debug enabled

rospogrigio committed:
Introduced device debugging option in README.md and info.md

rospogrigio committed:
Introduced IT and PT translations

rospogrigio committed:
Merge branch 'master' into device_debugging_option

rospogrigio committed:
Merge pull request #1227 from rospogrigio/device_debugging_option

Device debugging option

rospogrigio committed:
Update manifest.json for 5.0.0

rospogrigio committed:
New tuyadebug.tgz now supporting protocol 3.4

rospogrigio committed:
Update bug_report.md

rospogrigio committed:
Fix for unknown state of covers/shutters

rospogrigio committed:
New tuyadebug.tgz now including a tinytuya test

sibowler committed:
Merge branch 'master' into fix_semaphore_release_error

CloCkWeRX committed:
Force to int

CloCkWeRX committed:
Adjust input to avoid putting a string into an int

CloCkWeRX committed:
Adjust input to avoid putting a string into an int

CloCkWeRX committed:
Swap to int

rospogrigio committed:
Improved logging

rospogrigio committed:
Improved stability and Fix local_key update to not be thwarted by retries

rospogrigio committed:
Merge branch 'sibowler-fix_semaphore_release_error' into stability_fixes

rospogrigio committed:
Merge branch 'sibowler-fix_semaphore_release_error' into stability_fixes

rospogrigio committed:
Merge branch 'CloCkWeRX-fix-casting-issue' into stability_fixes

rospogrigio committed:
Merge branch 'stability_fixes' of github.com:rospogrigio/localtuya-homeassistant into stability_fixes

rospogrigio committed:
Introduced update of local_key when editing a device

rospogrigio committed:
Merge branch 'stability_fixes' of github.com:rospogrigio/localtuya into stability_fixes

rospogrigio committed:
Merge branch 'stability_fixes' of github.com:rospogrigio/localtuya into stability_fixes

rospogrigio committed:
Improved logging

rospogrigio committed:
Merge branch 'stability_fixes' of github.com:rospogrigio/localtuya-homeassistant into stability_fixes

rospogrigio committed:
Tox fixes

rospogrigio committed:
Tox fixes

rospogrigio committed:
Tox fixes

rospogrigio committed:
Fix for not reconnecting after 'giving up'

rospogrigio committed:
Tox fixes

rospogrigio committed:
Tox fixes, reduced to 3.9 only

MaximumSU committed:
sync to upstream

MaximumSU committed:
Sync to upstream

MaximumSU committed:
Merge branch 'master' of https://github.com/MaximumSU/localtuya

MaximumSU committed:
Update manifest.json

MaximumSU committed:
Upstream sync

X-Coder commented:
@MaximumSU Thank you for your work on these features which my device requires.
Today I tested your pull request: it worked very well without any problems. 
I tested base64 payloads, scaling, rounding, adding and editing of entities.

**Feedback / example setup:**
A little feedback of what I tried with my tuya rebranded smart inverter device BKW600-02 / GTB-600:
The device has two DPs which reports base64 encoded values (V|A|kW|hz) for inverter's input and output side.

I used for my smart inverter entity setup these setup:
DP 3 input (dc side): 
| base64 payload AVIARwAY | V | A | kW |
| -- | ------ | ----------- | --- |
|HEX|0152|0047|0018|
|DEC|338|71|24|
|**scaling**|0.1|0.1|0.01|
|**Value bytes range**|0:2|2:2|4:2|

DP 7 output (ac side):
|base64  payload CQwAA6oAANAB9A==| V | A | kW | Hz |
| --- | ------ | ----------- | --- | ----|
|HEX|0900|000365|0000c3|01f4|
|DEC|2304|869|195|500|
|**scaling**|0.1|0.001|0.001|0.1|
|**Value bytes range**|0:2|2:3|5:3|8:2|

I had only a very minor usability issue while typing in the 0 after the decimal separator in the scaling field: some requires 0.01 or 0.001 scaling. Typing in 0.0 deletes the whole field and set it to 0.
To overcome this, for 0.001 I had first to type 0.1 and then insert all the additional 0 between the decimal separator and last number.

MaximumSU commented:
> I had only a very minor usability issue while typing in the 0 after the decimal separator in the scaling field: some requires 0.01 or 0.001 scaling. Typing in 0.0 deletes the whole field and set it to 0.
> To overcome this, for 0.001 I had first to type 0.1 and then insert all the additional 0 between the decimal separator and last number.

Yes, i know about it. But this is the Home Assistant config flow issue. I can't control over it :-( I think, It is better to have the scaling as a divisor nor multiplier, but it break's the compatibility with existent setups. I see some work in config flow about this issue in the last updates of Hass, but not sure, it is finished, or not.

X-Coder commented:
Just found a bug if the local_key changes (for example by unlinking the device and re-adding)

I am able to update the local_key in the edit interface, it is stored, but existing device_config is not updated and HA is unable to fetch the existing entities afterwards.

In diagnostic information download file you can see my old and new local_key:
```json
"data": {
    "device_config": {
      "local_key": "123456",  // old key <-- this gets not updated
     },
     "device_cloud_info": {
      "local_key": "78910" // new key got updated
     }
}
```

MaximumSU commented:
> Just found a bug if the local_key changes

I did not any modifications in localkey handling...
As i see the code, it only detect the localkey change, when you open the "Edit a Device". It inserted the new localkey in "Local Key" field and write the notice neare "NOTE: a new local_key has been retrieved using cloud API" when you need to store this change. Is it not true?

X-Coder commented:
Thank you for your help, I tried it today again with "Edit a Device" but now the local_key change worked as it should, strange. Yesterday it was different, don't know why.

MaximumSU committed:
Reordering of strange exceptions order to achieving local key update

phrak commented:
Has this PR been integrated into the main project yet?

tmiranda8 commented:
I hope this PR will be integrated someday to the main branch, as to manage the base64 raw DPs are a pain in the ass

MaximumSU committed:
Update common.py

fix error 'async_dispatcher_send' called from a thread other than the event loop

BaltasarParreira commented:
Hi @MaximumSU will your forked version that allows same DP for various sensors be updated so it works with lasted 2025.1.2 CORE?
I had too roll back to the 2024.12.5 CORE since Localtuya was not working at all and using other Localtuya original or forked versions I can't use the same DP and base64 feature.

MaximumSU commented:
@BaltasarParreira 
> will your forked version that allows same DP for various sensors be updated so it works with lasted 2025.1.2 CORE?

I hope. Because i'm in the same boat ;-) But first i need upgrade to 2025 to see....

MaximumSU committed:
Home Assistant 2025.1 + compatibility

MaximumSU commented:
@BaltasarParreira 
It work now.

BaltasarParreira commented:
Hi, yesssss !!!
Great news and I confirm, all is working with the lasted CORE available.

Amazing work as always, thanks !!!

MaximumSU committed:
Update __init__.py

Deprecation fix: use async_register_admin_service

MaximumSU commented:
@BaltasarParreira  
> Hi, another core update and again looks like it breaks localtuya... arghghgh !!!
> 
> After update to 2025.5.0 I lost all entities.

Try the fix.

BaltasarParreira commented:
Hi, perfect all working again.
Mega !!!

Thanks.

BaltasarParreira commented:
HI, and we are back to another core update (2025.6.0  and 2025.6.1) that inoperates localtuya, ufffff !!!

MaximumSU commented:
@BaltasarParreira 
> HI, and we are back to another core update (2025.6.0 and 2025.6.1) that inoperates localtuya, ufffff !!!

How is error look's in log?

BaltasarParreira commented:
Ok quicker than I thought  ; )
I  just reinstall and I can't see anything special apart this kind:
2025-06-15 08:50:11.604 WARNING (MainThread) [custom_components.localtuya.common] [774...ff7] Failed to connect to 192.168.0.66: [Errno 113] Connect call failed ('192.168.0.66', 6668)

Is there anything else that can be done apart turn on "Enable debug logging" on the plugin?

BaltasarParreira commented:
It happens with all tuta devices, even normal switches and lamps.
the log has only that kind of lines, nothing else.  : (

BaltasarParreira commented:
@phrak  
Thanks, but that fork doesn't support the use of the same DP many times for sensors.
I have some Tuya devices that need this so I can get them working.

phrak commented:
Ok sure. I don't really understand the use case for having the same DP many times for sensors, but perhaps @MaximumSU would consider submitting their same PR for to @xZetsubou's excellent fork too?
https://github.com/xZetsubou/hass-localtuya

BaltasarParreira commented:
Ok, so I turn on the options you said and this is the only type of new lines showing:

2025-06-17 21:26:44.556 WARNING (MainThread) [custom_components.localtuya.pytuya] [bf9...gqt] Command 9 timed out waiting for sequence number -100


All the rest is the same as I posted here before.   : (

BaltasarParreira commented:
Hi, still no fix for using the latest core?
I tried so many things but none is working, only staying on the 2025.5.3 core localtuya works for all devices even the ones without the need for same DP on sensors.

MaximumSU committed:
Update common.py

fix error 'async_dispatcher_send' called from a thread other than the event loop

BaltasarParreira commented:
Ok I understand and thanks for trying spending your few time left helping, but is a no go.
None of the devices work and are identified.
I'll try another approach maybe the old way using external script for decoding the base64 devices using the normal localtuya instead of yours if that stills works also or last option I have o keep using the old core while still everything else and doesn't evolve depending of the new one.
  :  )

MaximumSU committed:
tuya 3.5  protocol support, improved cloud support

pituya interface module replaced, improved cloud support, added tuya 3.5 device protocol support.
